docker/build-tools/Dockerfile

```diff
@@ -120,9 +170,23 @@ ENV GOPATH=/go
 ENV GOCACHE=/gocache
 ENV PATH /go/bin:$PATH
 
+# Uhh ya. Ruby.
+ENV RUBYOPT="-KU -E utf-8:utf-8"
+
 COPY --from=base_os_context / /
 COPY --from=golang_context /usr/local/go/ /go/
+
+# The build_nodejs_context needs attention - probably need to build nodejs from source
+#COPY --from=build_nodejs_context / /
+
+COPY --from=build_hugo_context /out/ /
 COPY --from=build_tools_context /out/ /
+COPY --from=ruby_tools_context /usr/local/bin /usr/local/bin
+COPY --from=ruby_tools_context /usr/local/lib /usr/local/lib
+COPY --from=node_tools_context /usr/local /usr/local
+COPY --from=node_tools_context /package.json /
```

Comment:
Both the ruby and the nodejs stuff will end up carrying a whole bunch of stuff redundantly within the image I think. Can we be more selective in the context creation and but just the new Ruby- and Node-related bits in an /out directory so we end up with no redundant state in layers?

Either that, or just create one more scratch context at the end and copy from the previously final context into it. That'll flatten everything.

Reply:
I just tried this and it actually made things bigger.

Shows how much I know about this...

Reply:
I've spent a fair amount of time optimizing the build contexts - to 950mb.  Pretty good sizing I think.  In _addition_ we were using some fellas PPA in the website tools for node.  Now we are using the upstream release from an upstream source (as well as ruby).